flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/OperatorChain.java

```diff
@@ -106,7 +107,7 @@
 
 	public OperatorChain(
 			StreamTask<OUT, OP> containingTask,
-			List<RecordWriter<SerializationDelegate<StreamRecord<OUT>>>> recordWriters) {
+			RecordWriterDelegate<SerializationDelegate<StreamRecord<OUT>>> recordWriterDelegate) {
```

Comment:
Looks much nicer now. I wonder if we could also move SerializationDelegate type information inside of RecordWriterDelegate.

Reply:
If so it could make other references shorter. Let me try that.

Reply:
After double thinking, I have not found a proper way to make `SerializationDelegate` type inside `RecordWriterDelegate`. In other words, it would be more general as now to define generics `IOReadableWritable` in interface `RecordWriterDelegate`, to not limit the scope into `SerializationDelegate`.  Or are there any good suggestions to do so?

Reply:
Just an idea, not sure if this is too specific.

```java
public interface RecordWriterDelegate<OUT> extends AvailabilityProvider, AutoCloseable {
	RecordWriter<SerializationDelegate<StreamRecord<OUT>>> getRecordWriter(int outputIndex);
}
```
Then the signature of the Operator ctor would be
```java
	public OperatorChain(
			StreamTask<OUT, OP> containingTask,
			RecordWriterDelegate<OUT> recordWriters) {
```

Reply:
Yeah, the above way would make sense. 

But we have to migrate `RecordWriterDelegate` and the implementations into `flink-streaming-java` model because it would rely on the `StreamRecord` class. Now I put them into the `flink-runtime` model together with `RecordWriter` package. The existing `BatchTask` also uses `RecordWriter` instances, but i am not sure whether it would also use `RecordWriterDelegate` future. So from this point, it would no limitation to put it together with runtime model.

Another tiny consideration, the existing `RecordWriter` class also uses the `IOReadableWritable` generics, so the respective delegate keeps the same generics with the instances.